clients/src/test/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignorTest.java

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import static org.junit.Assert.assertTrue;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.kafka.clients.consumer.ConsumerPartitionAssignor.Subscription;
+import org.apache.kafka.clients.consumer.internals.AbstractStickyAssignor;
+import org.apache.kafka.clients.consumer.internals.AbstractStickyAssignorTest;
+import org.apache.kafka.common.TopicPartition;
+
+public class CooperativeStickyAssignorTest extends AbstractStickyAssignorTest {
+
+    @Override
+    public AbstractStickyAssignor createAssignor() {
+        return new CooperativeStickyAssignor();
+    }
+
+    @Override
+    public Subscription buildSubscription(List<String> topics, List<TopicPartition> partitions) {
+        return new Subscription(topics, assignor.subscriptionUserData(new HashSet<>(topics)), partitions);
+    }
+
+    /**
+     * The cooperative assignor must do some additional work and verification of some assignments relative to the eager
+     * assignor, since it may or may not need to trigger a second follow-up rebalance.
+     * <p>
+     * In addition to the validity requirements described in
+     * {@link org.apache.kafka.clients.consumer.internals.AbstractStickyAssignorTest#verifyValidityAndBalance(Map, Map, Map)},
+     * we must verify that no partition is being revoked and reassigned during the same rebalance. This means the initial
+     * assignment may be unbalanced, so if we do detect partitions being revoked we should trigger a second "rebalance"
+     * to get the final assignment and then verify that it is both valid and balanced.
+     */
+    @Override
+    public void verifyValidityAndBalance(Map<String, Subscription> subscriptions,
+                                         Map<String, List<TopicPartition>> assignments,
+                                         Map<String, Integer> partitionsPerTopic) {
+        int rebalances = 0;
+        // partitions are being revoked, we must go through another assignment to get the final state
+        while (verifyCooperativeValidity(subscriptions, assignments)) {
+
+            // update the subscriptions with the now owned partitions
+            for (Map.Entry<String, List<TopicPartition>> entry : assignments.entrySet()) {
+                String consumer = entry.getKey();
+                Subscription oldSubscription = subscriptions.get(consumer);
+                subscriptions.put(consumer, buildSubscription(oldSubscription.topics(), entry.getValue()));
+            }
+
+            assignments.clear();
+            assignments.putAll(assignor.assign(partitionsPerTopic, subscriptions));
+            ++rebalances;
+        }
+
+        assertTrue(rebalances <= 2);
```

Comment:
This assertion should be removed as well since we may go beyond two rebalances, right?

Reply:
To avoid the test runs indefinitely (say we did something wrong that triggers endless rebalances), maybe we can say fail after 10 `assignor.assign` calls?

Reply:
The assignment doesn't change randomly between these additional rebalances, so there's no reason the first follow up rebalance shouldn't get close enough to fully balanced for the second follow up rebalance to be the end. I've run it many times and two was always sufficient. If we're worried 2 will be flaky I suppose we could turn it up to 3, but that should absolutely be high enough. If the tests start needing 10 extra rebalances to reach a settled assignment, I would consider that a bug and not passing

Reply:
Cool, I mentioned this since you said it may be going for a third rebalance sometimes. If you are confident that 3, or even 2 now would be sufficient, then let's break the while loop and fail early too.

Reply:
Sorry, I think I've been interchangeability referring to this as 2 and 3 rebalances. What I meant is that there are never more than 2 extra rebalances (ie, never more than a third total rebalance) needed for this test.

Reply:
If we check that it should not be more than 2, then how about fail inside the while loop? If there's a bug causing `verifyCooperativeValidity` to always return false, it would run indefinitely.

Reply:
Good point, will move the check inside the loop